onnxruntime/core/optimizer/nchwc_transformer.cc

```diff
@@ -0,0 +1,742 @@
+                              const ONNX_NAMESPACE::TensorProto* filter_shape);
+
+  void TransformConv(Node& node);
+  void TransformPool(Node& node);
+  void TransformAdd(Node& node);
+  void TransformConcat(Node& node);
+  void TransformActivation(Node& node);
+  void TransformElementwise(Node& node);
+
+  Graph& graph_;
+
+  // Stores a queue of nodes to be removed after walking through the graph.
+  std::deque<NodeIndex> removed_nodes_;
+
+  // Stores a mapping from the original NodeArg outputs to the NCHWc variants
+  // created inside this graph transform.
+  std::unordered_map<NodeArg*, std::unique_ptr<NchwcArgument>> nchwc_args_;
+
+  // Stores a mapping of NodeArg inputs that have already been reordered, so
+  // multiple nodes can share the NCHWc input.
+  std::unordered_map<NodeArg*, NodeArg*> reorder_inputs_;
+
+  // Stores a mapping of NodeArg filters that have already been reordered, so
+  // multiple nodes can share the NCHWc filter.
+  std::unordered_map<NodeArg*, NodeArg*> filters_OIHWBo_;
+  std::unordered_map<NodeArg*, NodeArg*> filters_OIHWBiBo_;
+
+  // Stores a mapping of NodeArg biases that have already been aligned to the
+  // NCHWc block size, so multiple nodes can share the NCHWc biases.
+  std::unordered_map<NodeArg*, NodeArg*> aligned_biases_;
+};
+
+const ONNX_NAMESPACE::AttributeProto* NchwcTransformerImpl::GetAttribute(const Node& node, const char* attribute_name) {
+  auto& node_attributes = node.GetAttributes();
+  auto it = node_attributes.find(attribute_name);
+  if (it != node_attributes.end()) {
+    return &(it->second);
+  } else {
+    return nullptr;
+  }
+}
+
+size_t NchwcTransformerImpl::RemoveOutputEdges(Node& node) {
+  size_t output_edges_count = node.GetOutputEdgesCount();
+  if (output_edges_count > 0) {
+    std::vector<Node::EdgeEnd> output_edges;
+    output_edges.reserve(output_edges_count);
+    for (auto it = node.OutputEdgesBegin(); it != node.OutputEdgesEnd(); ++it) {
+      ORT_ENFORCE(it->GetSrcArgIndex() == 0);
+      output_edges.push_back(*it);
+    }
+    for (auto& edge : output_edges) {
+      graph_.RemoveEdge(node.Index(), edge.GetNode().Index(), edge.GetSrcArgIndex(), edge.GetDstArgIndex());
+    }
+  } else {
+    // Bias the edge count to handle the case of a node that produces a graph
+    // output.
+    output_edges_count = 1;
+  }
+  return output_edges_count;
+}
+
+void NchwcTransformerImpl::CreateNchwcArgument(Node& node,
+                                               Node& nchwc_node,
+                                               int64_t channels,
+                                               const NchwcArgument::Shape& shape) {
+  size_t original_uses = RemoveOutputEdges(node);
+
+  // Create a new NodeArg to track the output from the NCHWc node.
+  auto& output_defs = nchwc_node.MutableOutputDefs();
+  auto* output_original_arg = output_defs[0];
+  std::string output_reorder_def_name = graph_.GenerateNodeArgName("reorder");
```

Comment:
This works but it may be faster to just prepend or append reorder to the original name

Reply:
I would still need to call GenerateNodeArgName in order to guarantee that the name is unique across the graph (as unlikely there would be a collision). That's why I avoided generating a name from the original and just made something standalone.